src/storage/raw/ttl.rs

```diff
@@ -0,0 +1,384 @@
+
+use engine_traits::util::{get_expire_ts, strip_expire_ts, truncate_expire_ts};
+use engine_traits::CfName;
+use engine_traits::{IterOptions, ReadOptions};
+#[cfg(not(test))]
+use tikv_util::time::UnixSecs;
+use txn_types::{Key, Value};
+
+#[cfg(test)]
+pub const TEST_CURRENT_TS: u64 = 15;
+
+#[derive(Clone)]
+pub struct TTLSnapshot<S: Snapshot> {
+    s: S,
+    current_ts: u64,
+}
+
+impl<S: Snapshot> TTLSnapshot<S> {
+    fn map_value(&self, mut value_with_ttl: Result<Option<Value>>) -> Result<Option<Value>> {
+        if value_with_ttl.is_err() {
+            return value_with_ttl;
+        }
+
+        if let Some(v) = value_with_ttl.as_ref().unwrap().as_ref() {
+            let expire_ts = get_expire_ts(v)?;
+            if expire_ts != 0 && expire_ts <= self.current_ts {
+                return Ok(None);
+            }
+        }
+
+        if let Some(v) = value_with_ttl.as_mut().unwrap().as_mut() {
+            truncate_expire_ts(v).unwrap()
+        }
+        value_with_ttl
+    }
+
+    #[cfg(not(test))]
+    #[inline]
+    pub fn current_ts() -> u64 {
+        UnixSecs::now().into_inner()
+    }
+
+    #[cfg(test)]
+    #[inline]
+    pub fn current_ts() -> u64 {
+        TEST_CURRENT_TS
+    }
+
+    pub fn get_key_ttl_cf(
+        &self,
+        cf: CfName,
+        key: &Key,
+        stats: &mut Statistics,
+    ) -> Result<Option<u64>> {
+        let value_with_ttl = self.s.get_cf(cf, key);
+
+        if let Err(e) = value_with_ttl {
+            return Err(e);
+        }
+
+        stats.data.flow_stats.read_keys = 1;
+        stats.data.flow_stats.read_bytes = key.as_encoded().len();
+        if let Some(v) = value_with_ttl.as_ref().unwrap().as_ref() {
+            stats.data.flow_stats.read_bytes += v.len();
+            let expire_ts = get_expire_ts(v)?;
+            if expire_ts == 0 {
+                return Ok(Some(0));
+            }
+            return if expire_ts <= self.current_ts {
+                Ok(None)
+            } else {
+                Ok(Some(expire_ts - self.current_ts))
+            };
+        }
+        Ok(None)
+    }
+}
+
+impl<S: Snapshot> From<S> for TTLSnapshot<S> {
+    fn from(s: S) -> Self {
+        TTLSnapshot {
+            s,
+            current_ts: Self::current_ts(),
+        }
+    }
+}
+
+impl<S: Snapshot> Snapshot for TTLSnapshot<S> {
+    type Iter = TTLIterator<S::Iter>;
+
+    fn get(&self, key: &Key) -> Result<Option<Value>> {
+        self.map_value(self.s.get(key))
+    }
+
+    fn get_cf(&self, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.map_value(self.s.get_cf(cf, key))
+    }
+
+    fn get_cf_opt(&self, opts: ReadOptions, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.map_value(self.s.get_cf_opt(opts, cf, key))
+    }
+
+    fn iter(&self, iter_opt: IterOptions, mode: ScanMode) -> Result<Cursor<Self::Iter>> {
+        self.s
+            .iter(iter_opt, mode)
+            .map(|c| c.into_with(|i| TTLIterator::new(i, self.current_ts)))
```

Comment:
Confusing implementation. Why not make `S: engine_traits::Snapshot`?

Reply:
TTLSnapshot is used in the `storage` level, so what we can get from `async_snapshot()` is a `kv::Snapshot` instead of `engine_traits::Snapshot`.

Reply:
Why not add `ttl_get` to storage type directly?